pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,527 @@
+	"github.com/hashicorp/memberlist"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	crdinformers "antrea.io/antrea/pkg/client/informers/externalversions/crd/v1alpha2"
+	crdlister "antrea.io/antrea/pkg/client/listers/crd/v1alpha2"
+	"antrea.io/antrea/pkg/util/k8s"
+)
+
+const (
+	controllerName = "MemberListCluster"
+	// Set resyncPeriod to 0 to disable resyncing.
+	resyncPeriod time.Duration = 0
+	// set default virtual node replicas num of consistent hash
+	defaultVirtualNodeReplicas = 3
+)
+
+type localNodeEventType int
+
+const (
+	localNodeWillJoin localNodeEventType = iota
+	localNodeWillLeave
+)
+
+// default Hash Fn is crc32.ChecksumIEEE
+var defaultHashFn func(data []byte) uint32
+
+type ClusterInterface interface {
+	Run(stopCh <-chan struct{})
+	ShouldSelect(name string) bool
+	AddEventHandler(handler clusterNodeEventHandler)
+}
+
+type clusterNodeEventHandler func(nodeName string, isJoin bool)
+
+type Cluster struct {
+	bindPort int
+	// NodeConfig contains nodeName and ipAddr which would be used
+	NodeConfig *config.NodeConfig
+	// Name of local node. Node name must be unique in the cluster.
+	nodeName string
+
+	mList *memberlist.Memberlist
+	// conHash hold the consistenthash.Map, when a node join cluster, use method Add() to add some keys to the hash.
+	// when a node leave the cluster, the consistenthash.Map should update
+	conHash *consistenthash.Map
+	// 	nodeEventsCh, the Node join/leave events will be notified via it.
+	nodeEventsCh chan memberlist.NodeEvent
+
+	// ClusterNodeEventHandlers hold the fun that should implement when a node join or leave the cluster
+	// for example, when a nwe node join the cluster,
+	// each node should recalculate that if it should still hold the existing egresses owned by itself
+	// in the same way, when a node leave cluster, each node may get some new egresses from the left node
+	ClusterNodeEventHandlers []clusterNodeEventHandler
+
+	// existingMembers contains all node(ip:clusterPort) that should join memberlist cluster
+	// the items list from nodelist and filter with externalIPPool nodeSelectors
+	// if the local node is not joined Cluster, it should hold the nodes,
+	// once local node should join cluster, it will join cluster with those nodes
+	memberRWLock    sync.RWMutex
+	existingMembers sets.String
+
+	localNodeEventCh chan localNodeEventType
+
+	// nodeSelectors contains all the node selectors that list from externalIPPools
+	// used to compare node selector and calculate whether it should join cluster more efficiently
+	nodeSelectorRWLock sync.RWMutex
+	nodeSelectors      map[string]*metav1.LabelSelector
```

Comment:
This is just the nodeSelectors extracted from ExternalIPPool right? I think externalIPPoolLister itself is efficient to iterate and get an item (O(1) operation). Storing it in a separate map doesn't seem helping efficiency and need more code to maintain the map.

Reply:
done